src/dvc_data/index.py

```diff
@@ -238,8 +243,11 @@ def collect(index, path, fs):
 def save(index):
     from .hashfile.build import _build_file
 
-    for _, entry in index.iteritems():
+    dir_entries: List[DataIndexKey] = []
+
+    for key, entry in index.iteritems():
         if entry.meta.isdir:
+            dir_entries.append(key)
```

Comment:
Dir trees need to be built+saved last, after we've saved all the files that go inside the tree.

For worktree push/fetch we can technically save the tree without saving files since the index entries for pushed files already have hash_info/meta (from the serialized .dvc file). But for the general dir save case we won't always have the file entry hash_info/meta at this point (like for import-url)